### DISCUSSION TITLE
Implementation Guidance for Expand and Contract Pattern using Prisma Migrate

### DISCUSSION THREAD

**Discussion by hhimanshu** [prisma/prisma#21927] [Q&A] [answered]
### Question

Hello folks,

I have been reading about [Expand  and Contract Pattern](https://www.prisma.io/docs/guides/migrate/developing-with-prisma-migrate/customizing-migrations#example-use-the-expand-and-contract-pattern-to-evolve-the-schema-without-downtime) for evolving schema without downtime. I came across https://github.com/prisma/prisma/discussions/7421 and read thoroughly. What I am unable to understand is how many migrations and production deployments will be needed. Additionally, what goes in each migration.

I documented a flow that "I think" (based on my little understanding) make sense, but I am aware that my knowledge may have gaps. Could someone with experience with this pattern help me? Thanks in advance

The use case is **renaming the field**

#### Db Migration
1. Add a `new` field
#### Application
1. Update code to write to both `old` and the `new` field.
2. Read from the `old` field (No changes here)

> 🚀 Deploy changes in production 🚀 

#### DB Migration
1. Copy the data from `old` field to the `new` field
```
UPDATE "Profile" SET biography = bio;
```
Quoting [Step 3](https://www.prisma.io/docs/guides/migrate/developing-with-prisma-migrate/customizing-migrations#example-use-the-expand-and-contract-pattern-to-evolve-the-schema-without-downtime) from the docs

3. Mark `@ignore` on the `old` field.
#### Application
1. Stop writing to `old` field. Remove all references from code that "writes" to the new field
2. Start reading from the `new` field. By this time, the `new` field will be populated with *all* previous data from `old` column so it is safe reading from `new` field.

> 🚀 Deploy changes in production 🚀 

#### DB Migration
1. Remove `old` field from the schema

#### Application
1. Remove all references in the code that refer to `old` field.

> 🚀 Deploy changes in production 🚀 

So, as far as I see, it needs `3` production deployments for implementing **Expand and Contract** pattern. Am I missing anything?

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment, marked as answer]
Hi @hhimanshu 👋 

Thank you for raising this question.

Your understanding of the `Expand and Contract` pattern for evolving schema without downtime is correct. The process you've outlined is in line with the steps mentioned in the [Prisma documentation](https://www.prisma.io/docs/guides/migrate/developing-with-prisma-migrate/customizing-migrations#example-use-the-expand-and-contract-pattern-to-evolve-the-schema-without-downtime).

You're correct that it requires three production deployments to fully implement the `Expand and Contract` pattern when renaming a field. This approach helps to prevent downtime and data loss during schema evolution.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by hhimanshu** [reply]
Thank you for the confirmation @ludralph
